packages/number-input/src/NumberInput/NumberInput.tsx

```diff
@@ -0,0 +1,195 @@
+import PropTypes from 'prop-types';
+
+import { cx } from '@leafygreen-ui/emotion';
+import { useIdAllocator } from '@leafygreen-ui/hooks';
+import LeafyGreenProvider, {
+  useDarkMode,
+} from '@leafygreen-ui/leafygreen-provider';
+import { Description, Error, Label, Overline } from '@leafygreen-ui/typography';
+
+import { Input } from '../Input';
+import { Select } from '../Select';
+
+import {
+  errorMessageStyles,
+  errorMessageWrapperStyles,
+  labelDescriptionStyles,
+  unitBaseStyles,
+  unitDisabledStyles,
+  unitThemeStyles,
+  wrapperBaseStyles,
+  wrapperGapStyles,
+  wrapperSizeStyles,
+} from './NumberInput.styles';
+import { NumberInputProps, Size, State } from './NumberInput.types';
+
+export const NumberInput = React.forwardRef<HTMLInputElement, NumberInputProps>(
+  (
+    {
+      state = State.None,
+      size = Size.Default,
+      unitOptions = [],
+      onSelectChange = () => {},
+      darkMode: darkModeProp,
+      id: idProp,
+      'aria-describedby': ariaDescribedbyProp,
+      'aria-labelledby': ariaLabelledbyProp,
+      'aria-label': ariaLabelProp,
+      unit,
+      className,
+      disabled,
+      label,
+      value,
+      description,
+      errorMessage,
+      onChange,
+      popoverZIndex,
+      usePortal,
+      portalClassName,
+      portalContainer,
+      scrollContainer,
+      ...rest
+    }: NumberInputProps,
+    forwardedRef,
+  ) => {
+    const prefix = 'lg-numberinput';
+    const inputId = useIdAllocator({ prefix, id: idProp });
+    const errorMessageId = useIdAllocator({ prefix, id: ariaDescribedbyProp });
+    const descriptionId = useIdAllocator({ prefix });
+    const selectId = useIdAllocator({ prefix });
+    const { darkMode, theme } = useDarkMode(darkModeProp);
+
+    const hasUnit = !!unit;
+    const hasSelectOptions =
+      Array.isArray(unitOptions) && unitOptions.length > 1;
+
+    const renderUnitOnly = hasUnit && !hasSelectOptions;
+    const renderSelectOnly = hasUnit && hasSelectOptions;
+    const renderErrorMessage = state === State.Error && errorMessage;
+
+    const popoverProps = {
+      popoverZIndex,
+      usePortal,
+      portalClassName,
+      portalContainer,
+      scrollContainer,
+    } as const;
+
+    return (
+      <LeafyGreenProvider darkMode={darkMode}>
+        <div className={className}>
+          {(label || description) && (
+            <div className={labelDescriptionStyles}>
+              {label && (
+                <Label htmlFor={inputId} disabled={disabled}>
+                  {label}
+                </Label>
+              )}
+              {description && (
+                <Description id={descriptionId} disabled={disabled}>
+                  {description}
+                </Description>
+              )}
+            </div>
+          )}
+          <div
+            className={cx(wrapperBaseStyles, wrapperSizeStyles[size], {
+              [wrapperGapStyles]: renderUnitOnly,
+            })}
+          >
+            <Input
```

Comment:
does this component have the potential to live outside of NumberInput and be leveraged as a base elsewhere? Not blocking, but wondering if this was designed with that in mind?